tests/unit/s2n_security_policy_cert_preferences_test.c

```diff
@@ -91,6 +91,169 @@ int main(int argc, char **argv)
+        {
+            DEFER_CLEANUP(struct s2n_config *config = s2n_config_new(), s2n_config_ptr_free);
+            EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, "rfc9151"));
+
+            EXPECT_FAILURE_WITH_ERRNO(s2n_config_add_cert_chain_and_key_to_store(config, test_cert), S2N_ERR_SECURITY_POLICY_INCOMPATIBLE_CERT);
+
+            /* assert that no certs were loaded */
+            uint32_t domain_certs = 0;
+            EXPECT_EQUAL(s2n_config_get_num_default_certs(config), 0);
+            EXPECT_SUCCESS(s2n_map_size(config->domain_name_to_cert_map, &domain_certs));
+            EXPECT_EQUAL(domain_certs, 0);
+            EXPECT_EQUAL(s2n_config_get_num_default_certs(config), 0);
+        };
+
+        /* load a cert then configure an invalid security policy */
+        {
+            DEFER_CLEANUP(struct s2n_config *config = s2n_config_new(), s2n_config_ptr_free);
+            EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(config, test_cert));
+            const struct s2n_security_policy *default_sp = config->security_policy;
+            EXPECT_FAILURE_WITH_ERRNO(s2n_config_set_cipher_preferences(config, "rfc9151"), S2N_ERR_SECURITY_POLICY_INCOMPATIBLE_CERT);
+
+            /* assert that the security policy was not changed */
+            EXPECT_EQUAL(config->security_policy, default_sp);
+        };
+    };
+
+    /* s2n_connection cases */
+    {
+        /* setup a config with the default security policy and the test cert */
+        DEFER_CLEANUP(struct s2n_config *config = s2n_config_new(), s2n_config_ptr_free);
+        EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(config, test_cert));
+
+        /* set a config then set an invalid security policy override */
+        {
+            DEFER_CLEANUP(struct s2n_connection *conn = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            EXPECT_SUCCESS(s2n_connection_set_config(conn, config));
+            EXPECT_FAILURE_WITH_ERRNO(s2n_connection_set_cipher_preferences(conn, "rfc9151"), S2N_ERR_SECURITY_POLICY_INCOMPATIBLE_CERT);
+
+            /* assert that the security policy override was not successful */
+            EXPECT_NULL(conn->security_policy_override);
+        };
+        /* set a security_policy_override then set an invalid config */
+        {
+            DEFER_CLEANUP(struct s2n_connection *conn = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            EXPECT_SUCCESS(s2n_connection_set_cipher_preferences(conn, "rfc9151"));
+            struct s2n_config *default_config = conn->config;
+            EXPECT_FAILURE_WITH_ERRNO(s2n_connection_set_config(conn, config), S2N_ERR_SECURITY_POLICY_INCOMPATIBLE_CERT);
+
+            /* assert that the config was not changed */
+            EXPECT_EQUAL(conn->config, default_config);
+        };
+    };
+
+    /* certificate_signature_preferences_apply_locally behavior tests */
```

Comment:
You seem to be testing multiple functions per block here, it makes it really hard to read what's going on. I think it would be better to test all the behavior of one function that you want, and then move on to the next function.